sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedStatement.scala

```diff
@@ -36,8 +38,11 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 private[sql] abstract class ParsedStatement extends LogicalPlan {
   // Redact properties and options when parsed nodes are used by generic methods like toString
   override def productIterator: Iterator[Any] = super.productIterator.map {
-    case mapArg: Map[_, _] => conf.redactOptions(mapArg.asInstanceOf[Map[String, String]])
-    case other => other
+    case mapArg: Map[_, _] =>
+      // May match any Map type, e.g. Map[String, Int], due to type erasure
+      Try(conf.redactOptions(mapArg.asInstanceOf[Map[String, String]])).getOrElse(mapArg)
```

Comment:
Is this what you want?
```
    case mapArg: Map[String, String] => conf.redactOptions(mapArg)
    case other => other
```

Reply:
but, my question is should we still redact the string key when the value is not string, or redact the string value when the key is not string?

Reply:
It does not work due to type erasure. As a matter of fact, compiler gives this warning:
```
Warning:(41, 18) non-variable type argument String in type pattern
scala.collection.immutable.Map[String,String] (the underlying of Map[String,String]) is
unchecked since it is eliminated by erasure
           case mapArg: Map[String, String] =>
```

Reply:
Currently only `options` and `properties` need to be redacted, they are Map[String, String].

Reply:
ParsedStatement is an abstract class, we might not be able to know how it will be used in the future. We should try our best to avoid adding any assumption here. Could you try it? I think it is not a hard problem.

Reply:
I think the `Try` option is reasonable here. If any map is `Map[String, String]` then it will be redacted. Otherwise, it is passed on. Redaction only works on `Map[String, String]` so there is nothing to do for other maps.

Reply:
@gatorsmile `SQLConf.redactOptions` calls `Utils.redact` that only supports Seq[(String, String)]. We could enhance these Utils redact methods to support non-String keys or values, but that feels like a separate PR.

Reply:
In Spark source code, we always try to avoid rely on the exception handling if we can possibly avoid it. 

Also, we try our best to avoid make an assumption in the utility class. 

I think we can enhance these Utils redact methods in this PR.

Reply:
Another solution: let `redactOptions` take a map of any key and value type. Only evaluate and update keys/values that are actually strings. This would maybe make it more flexible and handle, say, a `Map[SomeEnum,String]` later.

Reply:
@dongjoon-hyun @gatorsmile @srowen Please review the new commit.

Please note the commit does not redact Map[String, non-String] because the replacement text is string. For non-String, e.g, Int, if we use -1 as replacement, it may be confusing.